Source/ARTLocalDevice.m

```diff
@@ -102,6 +95,48 @@ + (ARTLocalDevice *)load:(NSString *)clientId storage:(id<ARTDeviceStorage>)stor
     return device;
 }
 
+- (BOOL)setupDetailsWithClientId:(NSString *)clientId {
+    NSString *deviceId = self.id;
+    NSString *deviceSecret = self.secret;
+    
+    if ([deviceId isEqualToString:@""] || deviceSecret == nil) { // generate both at the same time
+        deviceId = [self.class generateId];
+        deviceSecret = [self.class generateSecret];
+        
+        [_storage setObject:deviceId forKey:ARTDeviceIdKey];
+        [_storage setSecret:deviceSecret forDevice:deviceId];
+    }
+    
+    self.id = deviceId;
+    self.secret = deviceSecret;
+    
+    NSString *localClientId = self.clientId;
+    
+    if (localClientId && clientId && ![localClientId isEqualToString:clientId]) {
+        ARTLogError(self.logger, @"ARTLocalDevice: `clientId`s don't match: %@ vs %@.", clientId, localClientId);
+        return NO;
+    }
+    if (clientId) {
+        self.clientId = clientId;
+    }
+    return YES;
+}
+
+- (void)resetDetails {
+    self.id = @"";
```

Comment:
Why it's empty string and not nil? Is empty string shows that id hasn't been reseted?